### PR TITLE
Activate VT-d by vmcall

### PR DESCRIPTION
**Pull request by YadongQi**
vmm: guest: add function to get guests list
vmm: fix wrong iteration of guests in template_tee
vmm: vtd: add vmcall to activate VT-d
vmm: vmx_timer: remove warning log
config: cic: enable VT-d with ACTIVATE_VTD_BY_VMCALL
config: cic: remove unnecessary flags

Signed-off-by: Qi, Yadong <yadong.qi@intel.com>

### CODE REVIEW

File: vmm/modules/vtd/vtd.c [hunk 1]
```diff
@@ -349,6 +359,26 @@ static void vtd_invalidate_cache(UNUSED guest_cpu_handle_t gcpu, void *pv)
...
+{
+	uint32_t i;
+	guest_handle_t guest;
+
+	if (!vtd_activated) {
+		/* Remove DMAR engine from Guest mapping */
+		for (i = 0; i < DMAR_MAX_ENGINE; i++) {
+			if (engine_list[i].reg_base_hpa) {
+				for (guest = get_guests_list(); guest != NULL; guest = guest->next_guest)
+					gpm_remove_mapping(guest, engine_list[i].reg_base_hpa, 4096);
```
Comment (James-Zzz) [inline vmm/modules/vtd/vtd.c:372]: Do we have PAGE_SIZE macro to replace this hard code number?

File: vmm/modules/vtd/vtd.c [hunk 2]
```diff
@@ -363,9 +393,14 @@ void vtd_init(void)
...
 
 	event_register(EVENT_GPM_SET, vtd_update_domain_mapping);
-	event_register(EVENT_GUEST_MODULE_INIT, vtd_guest_setup);
 	event_register(EVENT_RESUME_FROM_S3, vtd_reactivate_from_s3);
 	event_register(EVENT_GPM_INVALIDATE, vtd_invalidate_cache);
+#ifndef ACTIVATE_VTD_BY_VMCALL
+	event_register(EVENT_GUEST_MODULE_INIT, vtd_guest_setup);
+#endif
+
+#define VMCALL_ACTIVATE_VTD 0x56544400ULL        // "VTD"
```
Comment (James-Zzz) [inline vmm/modules/vtd/vtd.c:402]: move this macro definition to line 34 to keep all macros there?

File: vmm/modules/vtd/vtd.c [hunk 3]
```diff
@@ -383,10 +418,16 @@ static void vtd_enable_dmar(vtd_engine_t *engine, uint64_t rt_hpa)
...
+	vtd_send_global_cmd(engine, VTD_GCMD_SRTP, TRUE);
 
 	/* Translation Enable */
-	vtd_send_global_cmd(engine, VTD_GCMD_TE);
+	vtd_send_global_cmd(engine, VTD_GCMD_TE, TRUE);
+}
+
+static void vtd_disable_dmar(vtd_engine_t *engine)
+{
+	/* Translation disable */
```
Comment (James-Zzz) [inline vmm/modules/vtd/vtd.c:429]: To keep align with 423, "Translation Disable" ?

File: vmm/modules/vtd/vtd.c [hunk 4]
```diff
@@ -363,9 +393,14 @@ void vtd_init(void)
...
 	event_register(EVENT_GPM_SET, vtd_update_domain_mapping);
-	event_register(EVENT_GUEST_MODULE_INIT, vtd_guest_setup);
 	event_register(EVENT_RESUME_FROM_S3, vtd_reactivate_from_s3);
 	event_register(EVENT_GPM_INVALIDATE, vtd_invalidate_cache);
+#ifndef ACTIVATE_VTD_BY_VMCALL
+	event_register(EVENT_GUEST_MODULE_INIT, vtd_guest_setup);
+#endif
+
+#define VMCALL_ACTIVATE_VTD 0x56544400ULL        // "VTD"
+	vmcall_register(GUEST_REE, VMCALL_ACTIVATE_VTD, vmcall_activate_vtd);
```
Comment (kwang13) [inline vmm/modules/vtd/vtd.c:403]: register vmcall only when ACTIVATE_VTD_BY_VMCALL is set.
BTW, in this case, we can move the global variable "vtd_activated" into vmcall_activate_vtd(), as static local variable.

File: vmm/modules/vtd/vtd.c [hunk 5]
```diff
@@ -349,6 +359,26 @@ static void vtd_invalidate_cache(UNUSED guest_cpu_handle_t gcpu, void *pv)
 	}
 }
 
+static void vmcall_activate_vtd(guest_cpu_handle_t gcpu UNUSED)
+{
+	uint32_t i;
+	guest_handle_t guest;
+
+	if (!vtd_activated) {
```
Comment (GangSecurity) [inline vmm/modules/vtd/vtd.c:367]: suggest "if (vtd_activated) return;" to make code simple.

File: vmm/modules/vtd/vtd.c [hunk 6]
```diff
@@ -147,18 +152,23 @@ static void vtd_calculate_trans_cap(uint8_t *max_leaf, uint8_t *tm, uint8_t *sno
 		"VT-d: 3/4-level page-table is NOT supported!\n");
 }
 
-static void vtd_send_global_cmd(vtd_engine_t *engine, uint32_t cmd)
+static void vtd_send_global_cmd(vtd_engine_t *engine, VTD_GCMD_BIT_OFF off, boolean_t set)
```
Comment (kwang13) [inline vmm/modules/vtd/vtd.c:155]: rename:
VTD_GCMD_BIT_OFF -> VTD_GCMD_BIT. OFF is confusing. In spec, the name is "Bits". So just use BIT here.
off -> bit.
boolean_t set -> uint32_t value. by using "value", you will put 0/1 there, which is more readable than True/False

File: vmm/modules/vtd/vtd.c [hunk 7]
```diff
@@ -349,6 +359,26 @@ static void vtd_invalidate_cache(UNUSED guest_cpu_handle_t gcpu, void *pv)
 	}
 }
 
+static void vmcall_activate_vtd(guest_cpu_handle_t gcpu UNUSED)
```
Comment (kwang13) [inline vmm/modules/vtd/vtd.c:362]: should it be wrapped with ACTIVATE_VTD_BY_VMCALL? will it lead to "unused function" error when it is not compiled?

File: vmm/modules/vtd/vtd.c (hunk 2 above)
Comment (YadongQi) [reply vmm/modules/vtd/vtd.c:402]: I perfer keep it here.

File: vmm/modules/vtd/vtd.c [hunk 8]
```diff
@@ -20,6 +20,8 @@
 
 #include "lib/util.h"
 #include "modules/vtd.h"
+#include "modules/template_tee.h"
```
[the remark this answers is not in the document]
Comment (YadongQi) [reply vmm/modules/vtd/vtd.c:23]: In order to use **GUEST_REE** when do vmcall_register()

**Comment by YadongQi** [comment]
> We can replace CML. This is new BIOS behavior. We can say for new BIOS implementation.

Make sense. Done.

**Comment by James-Zzz** [comment]
LGTM

**Pull request merged**